### PR TITLE
added iterators for dfs and bfs

### PR DESCRIPTION
**Pull request by kylebeggs**
Adds support to iterate through the nodes in a graph in a depth-first or breadth-first manner. (see #106 )

### CODE REVIEW

**Comment by kylebeggs** [comment]
This is failing because of other parts of the code, not what is included in this PR. I'm still kinda new to this - what is the procedure when this happens?

**Comment by simonschoelly** [comment]
> The checks are failing because of other parts of the code, not what is included in this PR. I'm still kinda new to this - what is the procedure when this happens?

From the error logs, it looks as if in your PR you used some names that where already used for something else:
```
WARNING: using Traversals.tree in module Main conflicts with an existing identifier.
``

**Comment by kylebeggs** [comment]
Ah, I see. I did use the same identifier as in the Experimental module, but since the goal here is to bring this functionality forefront can I delete what is in Experimental that conflicts?

**Comment by simonschoelly** [comment]
> Ah, I see. I fixed the issue. Tests still show one error but it seems to be in a completely different module

Indeed - apparently the tests are failing in Julia v1.8 -> I will need to investigate here

**Comment by simonschoelly** [comment]
@kylebeggs If you rebase/merge the lastest master in your branch, the tests should pass now.

**Review by etiennedeg — COMMENTED** [review]
Thanks a lot for your work on this, very appreciated.

File: src/traversals/iterators.jl [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+    source::Int
+end
+
+
+"""
+    mutable struct GraphIteratorState
+
+`GraphIteratorState` is a struct to hold the current state of iteration which is need for Julia's Base.iterate() function.
+"""
+mutable struct GraphIteratorState
```
[the remark this answers is not in the document]
Comment (kylebeggs) [reply src/traversals/iterators.jl:66]: Sure, makes sense. I'll make the changes.

File: src/traversals/iterators.jl [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+
+
+"""
+    Base.iterate(t::IteratorAlgorithm)
+
+First iteration to visit each node.
+"""
+function Base.iterate(t::IteratorAlgorithm)
+    visited = falses(nv(t.graph))
+    visited[t.source] = true
```
[the remark this answers is not in the document]
Comment (kylebeggs) [reply src/traversals/iterators.jl:79]: No, we don't guarantee that. Good catch. I will think of how to fix this and please if you have any suggestions do say. Thanks

[the remark this answers is not in the document]
Comment (kylebeggs) [reply src/traversals/iterators.jl:87]: For now, I used a Union type. I'm sure its not the best way to do it, but its 'safe'. I was also thinking to make another abstract type of iterator that would encompass all iterators you need to supply a single source node to start. Something like

```Julia
abstract type SingleSourceIterator <: IteratorAlgorithm end
struct BFSIterator <: SingleSourceIterator
    graph::AbstractGraph
    source::Int
end
```

The issue is I'm not too well versed in graphs. I just need to use DFS for one part of my own package and am still getting up to speed with graph algorithms, so I'm not sure what are appropriate names for these types.

File: src/traversals/iterators.jl [hunk 3]
```diff
@@ -0,0 +1,130 @@
+"""
+    abstract type IteratorAlgorithm
+
+`IteratorAlgorithm` is an abstract type which specifies a particular algorithm to use when iterating through a graph.
+"""
+abstract type IteratorAlgorithm end
```
Comment (simonschoelly) [inline src/traversals/iterators.jl:6]:
```suggestion
abstract type VertexIterator end
```
I think this makes it more clear, that we iterate over vertices.

Comment (kylebeggs) [reply src/traversals/iterators.jl:6]: Agreed, I'll update and perhaps wait on thoughts on my reply above before making another commit to possible to include that as well. Like I said in the above reply, I'm not too well verse in graphs so this helps a lot. Thanks

Comment (simonschoelly) [reply src/traversals/iterators.jl:6]: Also, we might parametrize the Iterator in the type of vertex that we return:
```julia
abstract type VertexIterator{T <: Integer} end
```

File: src/traversals/iterators.jl [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+    source::Int
+end
+
+
+"""
+    abstract type AbstractGraphIteratorState
+
+`BasicGraphIteratorState` is an abstract type to hold the current state of iteration which is need for Julia's Base.iterate() function.
+"""
+abstract type AbstractGraphIteratorState end
```
Comment (simonschoelly) [inline src/traversals/iterators.jl:66]: Is there any advantage of declaring an abstract type here, especially when we only have one subtype?

File: src/traversals/iterators.jl [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+end
+
+
+"""
+    Base.iterate(t::DFSIterator, state::BasicGraphIteratorState)
+
+Iterator to visit each node in a depth-first manner.
+"""
+function Base.iterate(t::DFSIterator, state::BasicGraphIteratorState)
+    while !isempty(state.queue)
```
Comment (simonschoelly) [inline src/traversals/iterators.jl:99]: If I understand this correctly, if the graph is is not connected (or only weakly connected in the directed case), we might not visit all vertices? Is that on purpose?

Comment (etiennedeg) [reply src/traversals/iterators.jl:99]: It's a good catch but I think it make sense to only visit vertices reachable from the source, and we could have a separate Iterator (with no source) that would call a DFS Iterator on unexplored vertices until all vertices are explored.

File: src/traversals/iterators.jl (hunk 4 above)
Comment (etiennedeg) [reply src/traversals/iterators.jl:66]: Well, he applied my review. This leaves an API for Iterators that would store more information, for example current depth of the vertex, visited edges... I think there should be some work to define a clear API for all the types of iterators we would like to support.

Comment (kylebeggs) [reply src/traversals/iterators.jl:66]: Yeah I think we need an API because I'd also like to implement some iterators for weighted graphs to do minimum spanning trees for instance. I don't know the algorithms for that, but I'd imagine that would require some more information to be stored in the state.

File: src/traversals/iterators.jl (hunk 5 above)
Comment (kylebeggs) [reply src/traversals/iterators.jl:99]: The assumption when doing DFS or BFS is you only explore connected vertices, at least that's my understanding (and personal use case)

File: src/traversals/iterators.jl (hunk 3 above)
Comment (kylebeggs) [reply src/traversals/iterators.jl:6]: Sure. I didn't realize there are other types of vertices.. can you point me to an example so I can read up

**Comment by kylebeggs** [comment]
@simonschoelly @etiennedeg  I took a new approach to this taking into account some of the comments. Let me know what you think. I converted this PR to a draft to get some feedback before writing the appropriate tests.

File: src/iterators/bfs.jl [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+5
+3
+1
+4
+5
+2
+```
+"""
+struct BFSIterator{S} <: VertexIterator
+    graph::AbstractGraph
```
Comment (thchr) [inline src/iterators/bfs.jl:36]: The type of the graph should be concrete, i.e., change to:
```suggestion
struct BFSIterator{S, T<:AbstractGraph{S}} <: VertexIterator
    graph::T
```

**Comment by kylebeggs** [comment]
@thchr thanks for the suggestion. @simonschoelly @etiennedeg I am opening this back up as I have made some changes. Please review and lets all work together to come up with something to get us started with iterators in this package.

**Comment by Tortar** [comment]
This is very cool, one thing I want to suggest for the bfs part is that to apply the optimization I already applied here: https://github.com/JuliaGraphs/Graphs.jl/pull/250.

File: src/iterators/bfs.jl [hunk 7]
```diff
@@ -0,0 +1,61 @@
+"""
+    BFSIterator
+
+`BFSIterator` is a subtype of [`VertexIterator`](@ref) to iterate through graph vertices using a breadth-first search. A source node(s) is optionally supplied as an `Int` or an array-like type that can be indexed if supplying multiple sources. If no source is provided, it defaults to the first vertex.
```
Comment (Tortar) [inline src/iterators/bfs.jl:4]: Let me know if I'm wrong but the `BFSIterator` doesn't seem to allow multiple sources in the current form

File: src/iterators/bfs.jl [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+
+"""
+    Base.iterate(t::BFSIterator)
+
+First iteration to visit each vertex in a graph using breadth-first search.
+"""
+function Base.iterate(t::BFSIterator)
+    visited = falses(nv(t.graph))
+    visited[t.source] = true
+    return (t.source, VertexIteratorState(visited, [t.source]))
```
Comment (Tortar) [inline src/iterators/bfs.jl:40]: In particular this part doesn't seem to work with e.g. a vector of nodes

File: src/iterators/bfs.jl (hunk 7 above)
Comment (kylebeggs) [reply src/iterators/bfs.jl:4]: That is correct

**Comment by kylebeggs** [comment]
@Tortar Thanks for the comments. Unfortunately, I don't have time to apply your suggestions right now. Please, make the changes if you are able to. Hopefully one day we can merge this.

**Comment by Tortar** [comment]
@kylebeggs I'm willing to go on with this, I think we have two options: you can list me as a collaborator of your fork so that I can update this branch, otherwise I will need to superseed this pr. I would prefer the first option so that you appear as the main author

**Comment by kylebeggs** [comment]
@Tortar sounds great! I'll add you as a collaborator..

**Comment by gdalle** [comment]
@kylebeggs if we give up on multiple sources for now and focus on a single source, is this in a reviewable / mergeable state?

**Comment by Tortar** [comment]
Hi @gdalle working on extending to multiple sources at the moment, shouldn't be too hard, sorry for the delay, but I forgot about this

**Comment by Tortar** [comment]
okay, now multi source versions for bfs and dfs should work, I don't know the Kruskal algorithm so I didn't update it. Also, tests are failing because of jet complaining about the kruskal implementation

**Comment by Tortar** [comment]
my idea would be to put that algorithm in a separate PR and finish the work for the simpler cases of bfs and dfs, what do you think?

**Comment by gdalle** [comment]
I agree, let's do BFS and DFS first, they're much more widely used

**Comment by Tortar** [comment]
Or if someone could help with that it would be even better

**Comment by Tortar** [comment]
I removed the Kruskal implementation, if you can @kylebeggs I think you should open the new PR though since you made it

**Comment by kylebeggs** [comment]
@Tortar I'm sorry, I'm confused. Why do we need to make a new PR? This one reflects the removal of Kruskal?

**Comment by Tortar** [comment]
I removed it from here so that we can finish the work for bfs/dfs, so maybe not to forgot about the kruskal implementation here: https://github.com/JuliaGraphs/Graphs.jl/pull/163/commits/72c6840dfd75cedc1d1a3286986bd19f3e522b48 you can open up a new pr

**Comment by Tortar** [comment]
anyway @gdalle it should be ready for review!

**Comment by kylebeggs** [comment]
@Tortar I see. You have done a lot here, please feel free to make that PR.

**Comment by gdalle** [comment]
Thank you for the PR! I cleaned up the code a little bit and ensured type stability in the structs

File: src/iterators/bfs.jl (hunk 7 above)
Comment (gdalle) [reply src/iterators/bfs.jl:4]: Now it does

File: src/iterators/bfs.jl [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+
+"""
+    BFSVertexIteratorState
+
+`BFSVertexIteratorState` is a struct to hold the current state of iteration
+in BFS which is needed for the `Base.iterate()` function.
+"""
+mutable struct BFSVertexIteratorState
+    visited::BitVector
+    queue::Vector{Int}
```
Comment (gdalle) [inline src/iterators/bfs.jl:34]: Something from https://github.com/JuliaCollections/DataStructures.jl might be more appropriate

File: src/iterators/dfs.jl [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+
+"""
+    DFSVertexIteratorState
+
+`DFSVertexIteratorState` is a struct to hold the current state of iteration
+in DFS which is needed for the `Base.iterate()` function.
+"""
+mutable struct DFSVertexIteratorState
+    visited::BitVector
+    queue::Vector{Int}
```
Comment (gdalle) [inline src/iterators/dfs.jl:34]: Something from https://github.com/JuliaCollections/DataStructures.jl might be more appropriate

File: src/iterators/bfs.jl [hunk 11]
```diff
@@ -0,0 +1,95 @@
...
+    source::S
+end
+
+"""
+    BFSVertexIteratorState
+
+`BFSVertexIteratorState` is a struct to hold the current state of iteration
+in BFS which is needed for the `Base.iterate()` function.
+"""
+mutable struct BFSVertexIteratorState
```
Comment (gdalle) [inline src/iterators/bfs.jl:32]: Can you describe the fields of the struct in the docstring?

File: src/iterators/dfs.jl [hunk 12]
```diff
@@ -0,0 +1,83 @@
...
+    source::S
+end
+
+"""
+    DFSVertexIteratorState
+
+`DFSVertexIteratorState` is a struct to hold the current state of iteration
+in DFS which is needed for the `Base.iterate()` function.
+"""
+mutable struct DFSVertexIteratorState
```
Comment (gdalle) [inline src/iterators/dfs.jl:32]: Can you describe the fields of the struct in the docstring?

File: src/iterators/bfs.jl [hunk 13]
```diff
@@ -0,0 +1,95 @@
...
+    state = BFSVertexIteratorState(visited, copy(t.source), 1, 1)
+    return (first(t.source), state)
+end
+
+"""
+    Base.iterate(t::BFSIterator, state::VertexIteratorState)
+
+Iterator to visit each vertex in a graph using breadth-first search.
+"""
+function Base.iterate(t::BFSIterator, state::BFSVertexIteratorState)
```
Comment (gdalle) [inline src/iterators/bfs.jl:67]: Can you add some comments in this function to explain the steps and help with reviewing?

File: src/iterators/dfs.jl [hunk 14]
```diff
@@ -0,0 +1,83 @@
...
+    state = DFSVertexIteratorState(visited, source_rev)
+    return (last(source_rev), state)
+end
+
+"""
+    Base.iterate(t::DFSIterator, state::VertexIteratorState)
+
+Iterator to visit each vertex in a graph using depth-first search.
+"""
+function Base.iterate(t::DFSIterator, state::DFSVertexIteratorState)
```
Comment (gdalle) [inline src/iterators/dfs.jl:66]: Can you add some comments in this function to explain the steps and help with reviewing?

File: src/iterators/bfs.jl [hunk 15]
```diff
@@ -0,0 +1,95 @@
...
+mutable struct BFSVertexIteratorState
+    visited::BitVector
+    queue::Vector{Int}
+    neighbor_idx::Int
+    n_visited::Int
+end
+
+BFSIterator(g::AbstractGraph) = BFSIterator(g, first(vertices(g)))
+
+Base.length(t::BFSIterator) = nv(t.graph)
```
Comment (gdalle) [inline src/iterators/bfs.jl:41]: I added a known length but on second thought it may not work for unconnected graphs. In which case the termination criteria in the iteration also need to change

File: src/iterators/dfs.jl [hunk 16]
```diff
@@ -0,0 +1,83 @@
...
+in DFS which is needed for the `Base.iterate()` function.
+"""
+mutable struct DFSVertexIteratorState
+    visited::BitVector
+    queue::Vector{Int}
+end
+
+DFSIterator(g::AbstractGraph) = DFSIterator(g, first(vertices(g)))
+
+Base.length(t::DFSIterator) = nv(t.graph)
```
Comment (gdalle) [inline src/iterators/dfs.jl:39]: I added a known length but on second thought it may not work for unconnected graphs. In which case the termination criteria in the iteration also need to change

File: test/iterators/dfs.jl [hunk 17]
```diff
@@ -0,0 +1,44 @@
+@testset "DFSIterator" begin
```
Comment (gdalle) [inline test/iterators/dfs.jl:1]: It would be good to test edge cases: empty graph, graph with 2 connected components

File: test/iterators/bfs.jl [hunk 18]
```diff
@@ -0,0 +1,44 @@
+@testset "BFSIterator" begin
```
Comment (gdalle) [inline test/iterators/bfs.jl:1]: It would be good to test edge cases: empty graph, graph with 2 connected components

**Review by gdalle — CHANGES_REQUESTED** [review]

File: src/iterators/bfs.jl (hunk 9 above)
Comment (Tortar) [reply src/iterators/bfs.jl:34]: I don't think so, removing from the start and the end is O(1) in a Vector if I'm not mistaken

File: src/iterators/dfs.jl (hunk 10 above)
Comment (Tortar) [reply src/iterators/dfs.jl:34]: same

File: src/iterators/bfs.jl (hunk 15 above)
Comment (Tortar) [reply src/iterators/bfs.jl:41]: yes, we need to assume an unknown length in all cases I guess

File: src/iterators/dfs.jl (hunk 16 above)
Comment (Tortar) [reply src/iterators/dfs.jl:39]: same as before

File: test/iterators/dfs.jl (hunk 17 above)
Comment (Tortar) [reply test/iterators/dfs.jl:1]: graph with 2 connected component is already tested

Comment (Tortar) [reply test/iterators/dfs.jl:1]: but adding an empty graph seems good

**Comment by Tortar** [comment]
Thanks @gdalle for the help! But tests are now failing somehow

File: test/iterators/bfs.jl (hunk 18 above)
Comment (Tortar) [reply test/iterators/bfs.jl:1]: same

File: src/iterators/bfs.jl (hunk 9 above)
Comment (gdalle) [reply src/iterators/bfs.jl:34]: I'm assuming the devil is in the constants, otherwise we wouldn't need stacks and queues

File: src/iterators/bfs.jl [hunk 19]
```diff
@@ -0,0 +1,95 @@
...
+
+"""
+    Base.iterate(t::BFSIterator, state::VertexIteratorState)
+
+Iterator to visit each vertex in a graph using breadth-first search.
+"""
+function Base.iterate(t::BFSIterator, state::BFSVertexIteratorState)
+    graph, visited, queue = t.graph, state.visited, state.queue
+    while !isempty(queue)
+        if state.n_visited == nv(graph)
```
Comment (gdalle) [inline src/iterators/bfs.jl:70]: @Tortar this is the termination criterion that is wrong for unconnected graphs

Comment (Tortar) [reply src/iterators/bfs.jl:70]: mmh, why is it wrong for unconnected graphs? this terminates the iterator only if all the nodes have been visited, it is more of an optimization for connected graph than anything else

File: src/iterators/bfs.jl (hunk 9 above)
Comment (Tortar) [reply src/iterators/bfs.jl:34]: will look for it, but I still bet on Vectors :D

File: src/iterators/bfs.jl (hunk 15 above)
Comment (gdalle) [reply src/iterators/bfs.jl:41]: yeah you're right sorry

File: src/iterators/bfs.jl (hunk 9 above)
Comment (gdalle) [reply src/iterators/bfs.jl:34]: a fair assumption ^^

Comment (Tortar) [reply src/iterators/bfs.jl:34]: checked...I lost the bet :D

Comment (gdalle) [reply src/iterators/bfs.jl:34]: woops ^^

**Comment by gdalle** [comment]
tests are fixed now

**Comment by kylebeggs** [comment]
Great to see so much progress on this @Tortar @gdalle. What else is needed to get this merged? There has been a lot of activity, so it is a little difficult to see what exactly is left to do?

**Comment by gdalle** [comment]
Essentially address my unresolved comments in the review: adding more tests and some comments / docstrings describing the logic to make reviewing easier

File: src/iterators/bfs.jl (hunk 9 above)
Comment (Tortar) [reply src/iterators/bfs.jl:34]: actually I checked before only on some micro-benchmarks but then I tried in the specific case on a simple graph and I actually saw a regression of 40% so in the end I think we can keep Vectors for now

File: test/iterators/dfs.jl (hunk 17 above)
Comment (Tortar) [reply test/iterators/dfs.jl:1]: added it

File: src/iterators/bfs.jl (hunk 11 above)
Comment (Tortar) [reply src/iterators/bfs.jl:43]: I tried but maybe it can be improved

**Comment by Tortar** [comment]
I tried to address all review comments by @gdalle, probably docstrings can be improved but hopefully the rest should be okay

File: src/iterators/bfs.jl (hunk 9 above)
Comment (gdalle) [reply src/iterators/bfs.jl:45]: since this feature does not exist yet, it's better to have it sooner than to have a fully optimized version maybe never

Comment (gdalle) [reply src/iterators/bfs.jl:45]: so let's do it this way and we can improve later if necessary

**Review by gdalle — APPROVED** [review]
Thank you for the contribution!
There are things we could improve, like using the `eltype` of the graph instead of `Int` or testing on directed graphs, but this has been cooking for far too long and it is a needed feature.

**Comment by kylebeggs** [comment]
Huge thanks to @Tortar !!
